x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java

```diff
@@ -972,6 +970,92 @@ public void testMaybeRemoveRemoteIndicesFromRoleDescriptors() {
         assertThat(empty.toBytesRef(), equalTo(Authentication.maybeRemoveRemoteIndicesFromRoleDescriptors(empty).toBytesRef()));
     }
 
+    public void testCheckConsistency() {
+        @SuppressWarnings("checkstyle:linelength")
+        final Map<String, String> testCases = Map.ofEntries(
+            entry("Anonymous authentication cannot have realm type [realm_1]", "45XtAwADZm9vAAoAAAABAARub2RlB3JlYWxtXzEHcmVhbG1fMQAAAwA="),
```

Comment:
Agreed that this is quite tricky to test. I have a couple of alternatives, none of which I am convinced of. Still worth discussing I think:

**Make the constructor `(Subject, Subject, Type)` package-private**

This one is really simple and would make these tests a lot easier to read, debug, and maintain since we could simply call the construct and `expectThrows` there. It undermines the whole effort that went into looking down the constructor though.

**Static method to invoke private constructor**

Something like:

```java
static void invokeConstructor(Subject effectiveSubject, Subject authenticatingSubject, AuthenticationType type) {
    new Authentication(effectiveSubject, authenticatingSubject, type);
}
```

This for sure feels hacky but it again allows us to have much cleaner and straight-forward tests, while keeping the private constructor effectively locked down.

**More explicit `StreamInput` construction**

The current approach makes it completely opaque what the test inputs are, and also makes it quite difficult to add new test cases. Unless I'm missing something it essentially requires modifying the code to remove the assertion code, running it to serialize an invalid instance, and then copy&pasting that over into the test.

I'm wondering about something like:

```java
final BytesStreamOutput out = new BytesStreamOutput();
Authentication.AuthenticationSerializationHelper.writeUserTo(new AnonymousUser(Settings.EMPTY), out);
new RealmRef("realm_1", "realm_1", "node").writeTo(out);
out.writeBoolean(false);
out.writeVInt(Authentication.AuthenticationType.ANONYMOUS.ordinal());
Authentication.writeMetadata(out, Map.of());
expectThrows(..., () -> new Authentication(StreamInput.wrap(BytesReference.toBytes(out.bytes()))));
```

It creates its own maintenance burden of having to deal with the "leaked" internal details of serialization, but makes the tests a little easier to follow and maintain (esp. if we introduce a test util method that essentially has a `(Subject, Subject, Type)` signature.

I don't think any of these are ideal & I'm on board with taking your original approach. 

Personally, I'm leaning (slightly) towards the hacky but pragmatic `invokeConstructor`, just because otherwise the tests really feel quite mysterious.

Reply:
Thanks a lot for the great input here. I really appreciate you taking time to give suggestions. They are very useful!

I agree that the current way of testing is rather non-ideal. After some more thoughts, I took an approach (https://github.com/elastic/elasticsearch/pull/93894/commits/dab7dd3afcbb3466e253423054631ac960bdd98b) which is a slight variant of your 3rd suggestion. For both `encode` and `writeTo` method, I extracted the actual code logic into new static helper methods that has signature of `doEncode(Subject, Subject, Type)` and `doWriteTo(Subject, Subject, Type, Out)`. The original methods just delegates to the new static methods which are also available for tests. 

This allows us to write more readable tests and also avoid leaking internals or changing visibility of constructors. The static methods are only for writing data out. So they do _not_ offer an alternative approach to instantiate an Authentication object which I think it is an important plus.

I also moved the consistency test to its own test class to not overwhelm existing tests.

Reply:
This is great 👍 Thanks for iterating!